server/src/sc/server/network/ClientManager.kt

```diff
@@ -75,17 +72,22 @@ class ClientManager : Runnable, IClientListener {
      */
     @Throws(IOException::class)
     fun start() {
-        this.clientListener.start()
-        if(this.thread == null)
-            this.thread = ServiceManager.createService(this.javaClass.simpleName, this).apply { start() }
+        clientListener.start()
+        if(serviceThread == null)
+            serviceThread = ServiceManager.createService(javaClass.simpleName, this).apply { start() }
     }
 
     fun close() {
-        this.running = false
-        this.thread?.interrupt()
-        this.clientListener.close()
-        // Stop clients in reverse order to ensure all clients are stopped since they might be removed by onClientDisconnected
-        this.clients.indices.reversed().forEach { clients[it].stop() }
+        running = false
+        serviceThread?.interrupt()
+        clientListener.close()
+        while (clients.size > 0) {
+            try {
+                clients.removeAt(clients.lastIndex).stop()
+            } catch (ignored: ArrayIndexOutOfBoundsException) {
+                // Client was removed concurrently
+            }
+        }
```

Comment:
None of these are clean approaches... have you considered making a copy of this first and then stop them from the copy?

Reply:
yes, but the thing is, if something is already removed from the list, it doesn't need to be stopped anymore. So this is actually the most sensible approach I could think of by now, since there is no simple "give me any element from the list atomically" method.

Reply:
I don't know enough about java in conjunction with concurrency issues. I'll just assume it isn't prone to undefined behaviour and that if it fails it fails with an actual ArrayIndexOutOfBoundsException